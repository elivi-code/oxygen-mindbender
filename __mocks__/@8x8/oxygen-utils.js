export * from '@8x8/oxygen-utils';

const useFontSize = jest.fn(() => 16);

export { useFontSize };
