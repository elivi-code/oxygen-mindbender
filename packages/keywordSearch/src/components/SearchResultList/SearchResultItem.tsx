import React, { FunctionComponent } from 'react';
import { SearchResultItemType } from '../../types';
import ButtonWrapper from './styled/ButtonWrapper';

export interface SearchResultItemProps {
  id: string;
  testId: string;
  isActive: boolean;
  item: SearchResultItemType;
  onClick: (item: SearchResultItemType) => void;
}

export const SearchResultItem: FunctionComponent<SearchResultItemProps> = ({
  id,
  testId,
  isActive,
  item,
  onClick,
}) => {
  return (
    <li tabIndex={-1}>
      <ButtonWrapper
        id={id}
        isActive={isActive}
        isInfo={false}
        onClick={() => onClick(item)}
        tabIndex={-1}
        title={Object.values(item).join(' , ')}
        data-test-id={testId}
      >
        <span>{Object.values(item).join(' â€¢ ')}</span>
      </ButtonWrapper>
    </li>
  );
};

export default SearchResultItem;
