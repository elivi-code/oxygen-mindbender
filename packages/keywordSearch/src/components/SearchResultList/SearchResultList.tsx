import React, { FunctionComponent, useEffect, useRef } from 'react';
import {
  EventKey,
  useArrowKeyListIndex,
  useGlobalActionKeyHandler,
} from '@8x8/oxygen-utils';
import { getTestAttributes } from '@8x8/oxygen-config';
import NoResultsWrapper from './styled/NoResultsWrapper';
import CardListWrapper from './styled/CardListWrapper';
import ViewResultsWrapper from './styled/ViewResultsWrapper';
import Divider from './styled/Divider';
import { SearchResultItemType, SearchResultTranslations } from '../../types';
import SearchResultItem from './SearchResultItem';
import { SEARCH_RESULT_LIST_ID } from '../../utils/constants';
import ListWrapper from './styled/ListWrapper';
import ButtonWrapper from './styled/ButtonWrapper';
import ResultsTextWrapper from './styled/ResultsTextWrapper';

export interface SearchResultsProps {
  searchResults: SearchResultItemType[];
  searchValue: string;
  resultsCount: number;
  onClickListItem: (val: SearchResultItemType) => void;
  onChangeActiveItem: (index: number) => void;
  onEnterListItem: (index: number) => void;
  onSearchSelect: (val: string) => void;
  translations: SearchResultTranslations;
  testId: string;
}

const useArrowKeySearchResultList = (
  elementCount: number,
  searchValue: string,
) => {
  const { focusedElementIndex } = useArrowKeyListIndex({
    elementCount,
    initialHighlightedElementIndex: 0,
  });

  return !searchValue ? 0 : focusedElementIndex;
};

export const SearchResultList: FunctionComponent<SearchResultsProps> = ({
  searchResults = [],
  searchValue,
  resultsCount,
  onClickListItem,
  onChangeActiveItem,
  onEnterListItem,
  onSearchSelect,
  translations,
  testId,
}) => {
  const focusedElementIndex = useArrowKeySearchResultList(
    searchResults.length + 1,
    searchValue,
  );
  const focusedElIndexRef = useRef<number>(focusedElementIndex);

  useEffect(() => {
    focusedElIndexRef.current = focusedElementIndex;
    const focusedEl = document.getElementById(focusedElementIndex.toString());

    onChangeActiveItem(focusedElIndexRef.current);
    focusedEl?.scrollIntoView({ block: 'nearest' });
  }, [focusedElementIndex, onChangeActiveItem]);

  useGlobalActionKeyHandler({
    actionKeyHandler: () => onEnterListItem(focusedElIndexRef.current),
    fireEvents: [EventKey.Enter],
  });

  if (searchValue && !searchResults.length) {
    return (
      <CardListWrapper>
        <NoResultsWrapper>
          {translations.noResultsText}&nbsp;
          <strong>{`"${searchValue}"`}</strong>
        </NoResultsWrapper>
      </CardListWrapper>
    );
  }

  const viewAllResultsText = searchValue
    ? `${translations.viewAllResultsForText} "${searchValue}"`
    : `${translations.viewAllResultsText}`;

  return (
    <CardListWrapper>
      <ViewResultsWrapper>
        <ButtonWrapper
          {...getTestAttributes(`${testId}-view-all-results-button`)}
          isInfo
          id="0"
          isActive={focusedElementIndex === 0}
          onClick={() => onSearchSelect(searchValue)}
          tabIndex={-1}
        >
          <ResultsTextWrapper>
            <span title={viewAllResultsText}>{viewAllResultsText}</span>
            &nbsp;({resultsCount})
          </ResultsTextWrapper>
        </ButtonWrapper>
        {searchValue && <Divider />}
      </ViewResultsWrapper>
      {searchValue && (
        <ListWrapper id={SEARCH_RESULT_LIST_ID}>
          {searchResults.map((item, index) => (
            <SearchResultItem
              key={Object.values(item)[0]}
              testId={`${testId}-list-item-button`}
              id={(index + 1).toString()}
              isActive={focusedElementIndex === index + 1}
              item={item}
              onClick={onClickListItem}
            />
          ))}
        </ListWrapper>
      )}
    </CardListWrapper>
  );
};

export default SearchResultList;
