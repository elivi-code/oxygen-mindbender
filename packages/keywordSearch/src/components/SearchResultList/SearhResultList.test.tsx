import { create, ReactTestRenderer } from 'react-test-renderer';
import React from 'react';
import 'jest-styled-components';
import noop from 'lodash.noop';
import { OxygenProvider } from '@8x8/oxygen-config';
import SearchResultList, { SearchResultsProps } from './SearchResultList';

jest.mock('./SearchResultItem', () => 'SearchResultItem');
jest.mock('@8x8/oxygen-list', () => 'List');
jest.mock('@8x8/oxygen-button', () => 'Button');

describe('<SearchResultList>', () => {
  const setup = (props?: Partial<SearchResultsProps>): ReactTestRenderer =>
    create(
      <OxygenProvider>
        <SearchResultList
          testId="keyword-search"
          searchResults={[]}
          searchValue="mockValue"
          onClickListItem={noop}
          onEnterListItem={noop}
          onSearchSelect={noop}
          onChangeActiveItem={noop}
          resultsCount={0}
          translations={{
            viewAllResultsForText: 'View all results for',
            noResultsText: "We couldn't find any results for",
            viewAllResultsText: 'View all results',
            loadingSearchResultsText: 'Searching for',
          }}
          {...props}
        />
      </OxygenProvider>,
    );

  it('should render correctly when no results found', () => {
    const wrapper = setup().toJSON();

    expect(wrapper).toMatchSnapshot();
  });

  it('should render correctly when search results found', () => {
    const wrapper = setup({
      searchResults: [{ 0: 'mockName1' }, { 0: 'mockName2', 1: 'mockGroup' }],
      searchValue: 'mock',
      resultsCount: 10,
    }).toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
