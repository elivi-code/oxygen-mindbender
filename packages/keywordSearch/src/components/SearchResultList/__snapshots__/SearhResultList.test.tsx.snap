// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchResultList> should render correctly when no results found 1`] = `
.c2 {
  padding: 20px 16px 20px 32px;
  display: block;
  text-align: left;
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 400;
  -webkit-letter-spacing: -0.006rem;
  -moz-letter-spacing: -0.006rem;
  -ms-letter-spacing: -0.006rem;
  letter-spacing: -0.006rem;
  color: #666666;
}

.c0 {
  height: 252px;
  width: 328px;
  background: #FFFFFF;
  box-shadow: 0px 2px 6px 0px #29292940;
  border: 1px solid #E0E0E0;
  border-top: 1px solid #E0E0E0;
  border-radius: 0px;
  overflow: hidden;
  text-align: center;
}

.c1 {
  position: absolute;
  width: 100%;
  top: 100%;
  border: 1px solid #F1F1F1;
  box-sizing: border-box;
  height: auto;
  z-index: 100;
  box-shadow: 0px 2px 20px 0px #29292940;
  border-radius: 6px;
  margin-top: 8px;
}

<div
  className="c0 c1"
  data-test-id="CARD"
>
  <span
    className="c2"
  >
    We couldn't find any results for
     
    <strong>
      "mockValue"
    </strong>
  </span>
</div>
`;

exports[`<SearchResultList> should render correctly when search results found 1`] = `
.c0 {
  height: 252px;
  width: 328px;
  background: #FFFFFF;
  box-shadow: 0px 2px 6px 0px #29292940;
  border: 1px solid #E0E0E0;
  border-top: 1px solid #E0E0E0;
  border-radius: 0px;
  overflow: hidden;
  text-align: center;
}

.c1 {
  position: absolute;
  width: 100%;
  top: 100%;
  border: 1px solid #F1F1F1;
  box-sizing: border-box;
  height: auto;
  z-index: 100;
  box-shadow: 0px 2px 20px 0px #29292940;
  border-radius: 6px;
  margin-top: 8px;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 12px;
  padding: 12px 16px 12px 16px;
}

.c5 {
  display: inline-block;
  width: 100%;
  border-bottom: 1px #E0E0E0 solid;
}

.c6 {
  margin: 0;
  padding: 0 16px 12px 16px;
  list-style-type: none;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  white-space: nowrap;
  width: 100%;
  border: none;
  cursor: pointer;
  color: #0056E0;
  border-radius: 6px;
  padding: 8px 16px;
  background: #F1F1F1;
  font-family: Inter,sans-serif;
  font-size: 0.875rem;
  line-height: 1.25rem;
  font-weight: 400;
  -webkit-letter-spacing: -0.006rem;
  -moz-letter-spacing: -0.006rem;
  -ms-letter-spacing: -0.006rem;
  letter-spacing: -0.006rem;
}

.c3:hover {
  background: #F1F1F1;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 100%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c4 > span:first-of-type {
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div
  className="c0 c1"
  data-test-id="CARD"
>
  <div
    className="c2"
  >
    <button
      className="c3"
      data-id="keyword-search-view-all-results-button"
      data-test-id="keyword-search-view-all-results-button"
      id="0"
      onClick={[Function]}
      tabIndex={-1}
    >
      <span
        className="c4"
      >
        <span
          title="View all results for "mock""
        >
          View all results for "mock"
        </span>
         (
        10
        )
      </span>
    </button>
    <div
      className="c5"
    />
  </div>
  <ul
    className="c6"
    id="table-search-results"
  >
    <SearchResultItem
      id="1"
      isActive={false}
      item={
        {
          "0": "mockName1",
        }
      }
      onClick={[Function]}
      testId="keyword-search-list-item-button"
    />
    <SearchResultItem
      id="2"
      isActive={false}
      item={
        {
          "0": "mockName2",
          "1": "mockGroup",
        }
      }
      onClick={[Function]}
      testId="keyword-search-list-item-button"
    />
  </ul>
</div>
`;
